---
title: "Gene co-expression analysis reveals transcriptome divergence between wild and cultivated chickpea under drought stress"
author: "S Moenga"
date: "6/9/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)
```

*Physiological data Analysis*

```{r Figures 1, S2, S4}

#Trended transpiration analysis : Figure 1a

setwd("/Phys_data")

weights_info = read.csv("Molten_weights.csv", header =T)

library(dplyr)

df.summary <- weights_info %>%
  group_by(Genotype,group, ordered_variable,Water_treatment ) %>%
  summarise(
    sd = parameters::standard_error(molten_variables_weights, na.rm = TRUE),
    molten_variables_weights = mean(molten_variables_weights)
  )
df.summary

#plot
 ggplot(weights_info, 
        aes(x = ordered_variable,y = molten_variables_weights, color= Genotype))+
    geom_line(aes(group = group, linetype = Water_treatment), size=1,data = df.summary)+
 geom_errorbar(
    aes(ymin =  molten_variables_weights-sd, ymax =  molten_variables_weights+sd),
    data = df.summary, width = 0.3, size=0.7) +
  scale_linetype_manual(values=c("dashed", "solid", "twodash"))+
   scale_colour_colorblind()+
   theme(axis.text.x = element_text(color = "black",face = "bold", size = 12, angle = 60), axis.text.y = element_text(face = "bold", size = 12))+
   theme(legend.text = element_text(color = "black", size =12, face= "bold"))
 
 #################################################################################################
 
 #Used for generating cumulative transpiration,shoot biomass, weight per seed, seed count Figures 1b, S2, S4
 
 library(multcompView)
 
data=read.csv("Combined_data.csv", sep = ",", header = T, check.names = F )
head(data)

# Following assessment of assumptions of normality and heteroscedasticity  (not presented here); A two way anova (or one-way "grouped" anova as presented here):
model=lm( data$Seed_Count ~ data$Group )
ANOVA=aov(model)
 
# Tukey test to study each pair of treatment :
TUKEY <- TukeyHSD(x=ANOVA, 'data$Group', conf.level=0.95, na.rm = T)
 
# See Tukey test family-wise confidence representation :
plot(TUKEY , las=1 , col="brown")
 
 # group the treatments that are not different each other together.
generate_label_df <- function(TUKEY, variable){
 
     # Extract labels and factor levels from Tukey post-hoc 
     Tukey.levels <- TUKEY[[variable]][,4]
     Tukey.labels <- data.frame(multcompLetters(Tukey.levels)['Letters'])
     
     #I need to put the labels in the same order as in the boxplot :
     Tukey.labels$treatment=rownames(Tukey.labels)
     Tukey.labels=Tukey.labels[order(Tukey.labels$treatment) , ]
     return(Tukey.labels)
     }
 
# Apply the function on my dataset
LABELS <- generate_label_df(TUKEY , "data$Group")
 
# A panel of colors so that each treatment group would have the same label color :
my_colors = rep(c("darkorchid1","dimgrey","darkgoldenrod4"), 4)  ##colorblind friendly colors
 
# Draw the basic boxplot
a <- boxplot(data$Seed_Count  ~ data$Group , ylim=c(min(data$WUE , na.rm = T) , 1.1*max(data$Seed_Count, na.rm = T)) , outline = T, col=my_colors,width = rep(c(8,8,8),4),  ylab="Seed Count" , main=" Seed Count")
 
# add the letter over each box at a defined position
over <- 0.03*max( a$stats[nrow(a$stats),] )
 
#Add the labels
text( c(1:nlevels(data$Group)) , a$stats[nrow(a$stats),]+over , LABELS[,1]  , col=my_colors , cex = 1.2,)

#######################################################################################################

#For generating Figure 1c, d

  library("FactoMineR")
 
  #Impute any missing values
 library(missMDA)

Clean_PCA = imputeFAMD(data, ncp = 2)
  head(Clean_PCA$completeObs)
 
PCA_pca = PCA(
Clean_PCA$tab.disj[,9:16], scale.unit = TRUE, ncp = 10, graph = TRUE)
  ##Clean_PCA$tab.disj  
    
drought.pca <- prcomp(Clean_PCA$completeObs[,6:12], scale. =T)

#GENOTYPE_HIGHLIGHTED

#Here, grouping is by treatment ; if  by genotype, then "group" is changed to genotype and colors increased to 4

ggbiplot(drought.pca,choices = c(1,2) ,scale = 1, obs.scale = 1,var.scale = 0.5,pc.biplot = T,ellipse.prob = 0.50,groups =data$Water_Level, ellipse = T, circle = F, varname.abbrev = F,varname.size =3.7,var.axes=T,labels.size =20)+
  geom_point(size=3, aes(color = factor(data$Water_Level)))+
  scale_colour_manual(name="Genotype", values= c("darkorchid1","dimgrey","darkgoldenrod4"))+
  theme(aspect.ratio = 1.5)+
theme(legend.title = element_text(color = "black", size =12, face= "bold")) +
  theme(axis.text.x = element_text(face = "bold", size = 14), axis.text.y = element_text(face = "bold", size = 14))+
theme(legend.text = element_text(color = "black", size =10, face= "bold"))+
  
   guides(color = guide_legend(override.aes = list(size=2.5)))

# extract correlation coefficients and importance of components
drought.pca$rotation
summary(drought.pca)

#######################################################################################################

```

*TRANSCRIPTOME ANALYSIS, adapted from Law et al., 2016*

```{r Read summarization}
library('rtracklayer')
library(Rsubread)
library("org.Carietinum.eg.db")

bam.files <- list.files(path = "/Bam/Mapped", pattern = ".sorted.bam$", full.names = T)

features_count= featureCounts(files = bam.files, 
                              annot.ext= "/Ref_files/Ref.gff",
                              isGTFAnnotationFile=TRUE, 
                              strandSpecific = 2,
                              GTF.featureType = "gene",
                              GTF.attrType = "Dbxref",
                              useMetaFeatures=T, countMultiMappingReads=T, 
                              autosort= T, nthreads=4, allowMultiOverlap=T ) 
#write to disk
write.csv(features_count, file = "features_count.csv")

```


```{r Differential Expression analyis}
library(edgeR)
library(limma)

counts = read.csv("/Counts/features_count.csv", sep = ",", header=T,  row.names = 1)
str(counts)

metadata = read.table("/Counts/metadata.csv", sep = ",", header = T)

#generate DGE object
y= DGEList(counts) 
colnames(y)=metadata$Sample_name #rename column headers as needed; first check that metadata file sample order matches counts file order 
y$samples$Genotype <- metadata$Genotype
y$samples$Species <- metadata$Species
y$samples$Treatment <- metadata$Treatment
y$samples$group <- metadata$group

#low count filtering

#how many genes are o-expressed?
table(rowSums(y$counts==0)==42)

keep <- rowSums(cpm(y)>1) >= 3

y <- y[keep, , keep.lib.sizes=FALSE] #left with 20035 genes 

#calculate normalization factors, using Trimmed Means of M values (TMM), that will be later incorporated in DGE analysis
y <- calcNormFactors(y, method = "TMM")

## Global expression vizualization:

# Generating MDS plot with log CPM values
col.treat  = c("darkorchid1","yellowgreen","tomato4")
col.treat <- col.treat[match(y$samples$Treatment,
                                  levels(y$samples$Treatment))]
pch1=c(15,23,19,17)
geno_points <- c(15,23,19,17)[match(y$samples$Genotype, levels(y$samples$Genotype))]

plotMDS(cpm(y, log=T), gene.selection = "common", top =21000, dim.plot= c(1,2),cex =2,
col = col.treat, pch = geno_points)
legend("topleft", lwd=4, col = c("darkorchid1","yellowgreen","tomato4"),cex=1.3,
       legend = levels(y$samples$Treatment))
legend("top", pch = pch1,cex=1, 
       legend = levels(y$samples$Genotype))


## DGE Analysis
#Fit a linear model to the data with, assuming normal distribution

design1 <- model.matrix(~0+y$samples$group)
  colnames(design1) <- levels(group)

# Pre-designate contrasts for differential analysis
my.contrasts <- makeContrasts(
     ICC14778_Mild_vs_Control = ICC_14778.Mild-ICC_14778.Control,                       
  ICC14778_Severe_vs_Control =ICC_14778.Severe-ICC_14778.Control,
  ICC14778_Severe_vs_Mild =ICC_14778.Severe-ICC_14778.Mild,
  ICC8058_Mild_vs_Control =ICC_8058.Mild-ICC_8058.Control,
  ICC8058_Severe_vs_Control =ICC_8058.Severe-ICC_8058.Control,
  ICC8058_Severe_vs_Mild =ICC_8058.Severe-ICC_8058.Mild,
  Savur63_Mild_vs_Control =Savur_063.Mild-Savur_063.Control,
  Savur63_Severe_vs_Control =Savur_063.Severe-Savur_063.Control,
  Savur63_Severe_vs_Mild =Savur_063.Severe-Savur_063.Mild,
  Kalkan64_Mild_vs_Control =Kalkan_064.Mild-Kalkan_064.Control,
  Kalkan64_Severe_vs_Control =Kalkan_064.Severe-Kalkan_064.Control,
  Kalkan64_Severe_vs_Mild =Kalkan_064.Severe-Kalkan_064.Mild,
  levels = design1
)

# for species level differentiation:
design2 <- model.matrix(~0+group_species)
  colnames(design2) <- levels(group_species)
design2

my.contasts_per_species= makeContrasts(
    Ari_vs_Ret_control = Carietinum.Control - Creticulatum.Control,           
   Ari_vs_Ret_severe =Carietinum.Severe - Creticulatum.Severe,
   Ari_vs_Ret_mild = Carietinum.Mild-Creticulatum.Mild, 
  levels = design2)

### remove heteroscedasticity
y_voom <- voom(y, design1, plot=T)

#Remodel in limma 
vfit <- lmFit(y_voom, design1)
vfit1 <- contrasts.fit(vfit, contrasts=my.contrasts_2016)
efit <- eBayes(vfit1)


#extract Differentially expressed genes, by using the relevant coeficient (as in the contrast matrix)
topTable(efit, coef=2,resort.by="logFC", adjust.method = "BH", p.value=0.05, number =Inf)

```


```{r Clustering}
library(ComplexHeatmap) #(Gu et al., 2016)
library(dendextend)

#convert file with gene tags and LFC into matrix
drought_exp_all= as.matrix(drought_exp_all)
    
#set color scale
col_fun = colorRamp2(c(-2, 0, 2), c("green", "black", "red"))
col_fun(seq(-2, 2))

#Generate dendogram, default dist option is euclidean
dend_1= as.dendrogram(hclust(dist(drought_exp_all)),edgePar = c(cex = 1, lty = 1:3), horiz = T)
#Add distinguising color to clusters on dendogram 
dend = color_branches(dend_1, k = 10)

#generate heatmap
all_heat_dend = Heatmap(drought_exp_all, name = "L2FC", col= col_fun, show_row_names = F,use_raster = T, jitter =T, cluster_rows = dend, row_split = 10, row_dend_width = unit(70, "mm"))

```


```{r Patterned Differential expression analysis}

library(NOISeq)
library(maSigPro)

#Load an arrayed version of metadata file
edesign_abiotic = read.csv2("/Counts/edesign_table.csv", sep = ",") #make sure rownames are unique

#organize data , order the dataframes of both the array and the counts 
edesign_abiotic[,1] = as.numeric(as.character(edesign_abiotic[,1]))
rownames(edesign_abiotic)<- as.factor(rownames(edesign_abiotic))
edesign_abiotic = as.matrix(edesign_abiotic)

#Create a regression matrix for the full regression mode
design <- make.design.matrix(edesign_abiotic, degree = 2, time.col = 1, 
                   repl.col = 2, group.cols = c(3:ncol(edesign_abiotic)))

##calculate theta
    estimateDisp(tmm(y$counts), design1)

#compute a regression fit for each gene, with significant co-expression called at adjusted p.value 0f 0.05 (see q value)
fit_all <- p.vector(tmm_normalized_y, design = design, MT.adjust = "BH", Q = 0.05, counts = T, min.obs = 20, theta = 11.084)

#find significant variables for each gene
tstep_all = T.fit(fit_all)

#Extract list of significant genes , using an R2 threshold of 70% 
sigs_all = get.siggenes(tstep=tstep_all,
        rsq=0.70, vars="all")

#generate plots of clusters patterned expressio
see.genes(sigs_all$sig.genes,sigs_all$sig.genes$edesign,distance = "cor", min.obs =20, step.method = "two.ways.backward", cluster.method = "hclust",show.fit=F, k=15, alfa=0.05, color.mode= "rainbow", cexlab = 14, legend = T,summary.mode = "median", newX11 = F)


```


```{r Gene Set enrichment with GAGE and Pathview}
##Gene set enrichment
library(gage) # (Luo et al., 2009)

#Load the DEG list, in this case >1 LFC
Gene_List= topTable(efit, n=Inf, coef=2,sort.by = "logFC", resort.by="p")

Gene_List_fc= Gene_List$logFC
  length(Savur_fc)

names(Gene_List_fc) = Gene_List$ENTREZID

data(kegg.gs) #Invoke KEGG database

cam=kegg.gsets("cam") #Load C. arietinum kegg pathways

kg.cam =cam$kg.sets

fc.kegg.p <- gage(Savur.fc, gsets = kg.cam, ref = NULL, samp = NULL, same.dir=T,set.size = c(2, 500), FDR.adj= T, use.fold =T) 

#visuaLize with Pathview
out.suffix = "Tolerant_Severe_test_limma"

sel <- fc.kegg.p$greater[, "q.val"] < 0.1&
         !is.na(fc.kegg.p$greater[, "q.val"])
path.ids <- rownames(fc.kegg.p$greater)[sel]

sel.1 <- fc.kegg.p$less[, "q.val"] < 0.1 &
           !is.na(fc.kegg.p$less[,"q.val"])

path.ids.1 <- rownames(fc.kegg.p$less)[sel.1]

path.ids2 <- substr(c(path.ids, path.ids.1), 1, 8)

library(pathview)
#native kegg network view
 pv.out.list <- sapply(path.ids2, function(pid) pathview(
                       gene.data =Gene_List_fc, pathway.id = pid,
                      species = "cam", out.suffix=out.suffix))

```

